consensus/polybft/validators_snapshot.go

```diff
@@ -214,29 +221,96 @@ func (v *validatorsSnapshotCache) storeSnapshot(block uint64, snapshot AccountSe
 // Cleanup cleans the validators cache in memory and db
 func (v *validatorsSnapshotCache) cleanup() error {
 	if len(v.snapshots) >= validatorSnapshotLimit {
-		latestBlock := uint64(0)
+		latestEpoch := uint64(0)
 
 		for e := range v.snapshots {
-			if e > latestBlock {
-				latestBlock = e
+			if e > latestEpoch {
+				latestEpoch = e
 			}
 		}
 
-		block := latestBlock
-		cache := make(map[uint64]AccountSet, numberOfSnapshotsToLeaveInMemory)
+		startEpoch := latestEpoch
+		cache := make(map[uint64]*validatorSnapshot, numberOfSnapshotsToLeaveInMemory)
 
 		for i := 0; i < numberOfSnapshotsToLeaveInMemory; i++ {
-			if snapshot, exists := v.snapshots[block]; exists {
-				cache[block] = snapshot
+			if snapshot, exists := v.snapshots[startEpoch]; exists {
+				cache[startEpoch] = snapshot
 			}
 
-			block--
+			startEpoch--
 		}
 
 		v.snapshots = cache
 
-		return v.state.cleanValidatorSnapshotsFromDB(latestBlock)
+		return v.state.cleanValidatorSnapshotsFromDB(latestEpoch)
 	}
 
 	return nil
 }
+
+// getLastCachedSnapshot gets the latest snapshot cached
+// If it doesn't have snapshot cached for desired epoch, it will return the latest one it has
+func (v *validatorsSnapshotCache) getLastCachedSnapshot(currentEpoch uint64) (*validatorSnapshot, error) {
+	snapshot := v.snapshots[currentEpoch]
+	if snapshot != nil {
+		v.logger.Trace("Found snapshot in memory cache", "Epoch", currentEpoch)
+
+		return snapshot, nil
+	}
+
+	snapshot, err := v.state.getLastSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	return snapshot, nil
+}
+
+// getNextEpochEndingBlock gets the epoch ending block of a newer epoch
+// It start checking the blocks from the provided epoch ending block of the previous epoch
+func (v *validatorsSnapshotCache) getNextEpochEndingBlock(latestEpochEndingBlock uint64) (uint64, error) {
+	blockNumber := latestEpochEndingBlock + 1 // get next block
+
+	_, extra, err := getBlockData(blockNumber, v.blockchain)
+	if err != nil {
+		return 0, err
+	}
+
+	startEpoch := extra.Checkpoint.EpochNumber
+	epoch := startEpoch
+
+	for startEpoch == epoch {
+		blockNumber++
+
+		_, extra, err = getBlockData(blockNumber, v.blockchain)
+		if err != nil {
+			if errors.Is(err, blockchain.ErrNoBlock) {
+				return blockNumber - 1, nil
+			}
+
+			return 0, err
+		}
+
+		epoch = extra.Checkpoint.EpochNumber
+	}
```

Comment:
What do you think about altering this for loop so it iterates through blocks and checks is it epoch ending block using this function `isEpochEndingBlock` instead?
Asking because this part looks similar to the logic implemented in the `isEpochEndingBlock` function.

```go
_, extra, err = getBlockData(blockNumber, v.blockchain)
if err != nil {
	if errors.Is(err, blockchain.ErrNoBlock) {
		return blockNumber - 1, nil
	}

	return 0, err
}
```

Reply:
We could reuse it here, but it will mean that we would retrieve some blocks 2 times.

For example, you are on block 10 (which is the epoch ending block of epoch 1). You are searching for the epoch ending block for epoch 2, so you start from block 11. You get the header and extra for block 11, and call the `isEpochEndingBlock`. `isEpochEndingBlock` will get the block 12 to check if epoch on block 11 and 12 are different. If they are not, you return to the main loop, and get the extra and header for block 12, and call `isEpochEndingBlock` with its data, which will then get the data for block 13, to check if epochs are different, etc. That means some blocks are retrieved twice.

Yes, I can return the extra and header of the next block retrieved in `isEpochEndingBlock`, but that will that function to complex, and the plan is to reuse it (make it a helper function).